src/box/gc.c

```diff
@@ -744,6 +822,411 @@ gc_consumer_update(struct gc_consumer *consumer, const struct vclock *vclock)
 		gc_schedule_cleanup();
 }
 
+void
+gc_consumer_update(struct gc_consumer *consumer, const struct vclock *vclock)
+{
+	if (consumer->is_persistent) {
+		vclock_copy(&consumer->volatile_vclock, vclock);
+		/*
+		 * Do not wake up the background fiber if it's persisting
+		 * consumers right now because it may be waiting for a cbus
+		 * message, which doesn't tolerate spurious wakeups.
+		 */
+		if (gc.persist_scheduled++ == gc.persist_completed)
+			fiber_wakeup(gc.persist_fiber);
+	} else {
+		gc_consumer_update_impl(consumer, vclock);
+	}
+}
+
+/**
+ * Write the consumer to persistent WAL GC state.
+ */
+static int
+gc_consumer_persist(struct gc_consumer *consumer)
+{
+	if (!gc_has_persistent_consumers())
+		return 0;
+
+	char tuple_buf[UUID_STR_LEN + VCLOCK_STR_LEN_MAX + GC_NAME_MAX + 30];
+	char *tuple_end = tuple_buf;
+	tuple_end = mp_encode_array(tuple_end, 3);
+	tuple_end = mp_encode_str0(tuple_end, tt_uuid_str(&consumer->uuid));
+	tuple_end = mp_encode_vclock_ignore0(tuple_end, &consumer->vclock);
+	tuple_end = mp_encode_map(tuple_end, 1);
+	tuple_end = mp_encode_str0(tuple_end, "object");
+	tuple_end = mp_encode_str0(tuple_end, consumer->object_name);
+	assert((unsigned long)(tuple_end - tuple_buf) < sizeof(tuple_buf));
+
+	return box_replace(BOX_GC_CONSUMERS_ID, tuple_buf, tuple_end, NULL);
+}
+
+int
+gc_consumer_sync(struct gc_consumer *consumer)
+{
+	if (consumer->is_persistent) {
+		uint64_t scheduled = gc.persist_scheduled;
+		/*
+		 * Wait for two failures: one for in-progress update, if any,
+		 * and the second for update with actual volatile vclock of
+		 * the consumer.
+		 */
+		uint64_t failed = gc.persist_failed + 2;
+		while (gc.persist_completed < scheduled && gc.persist_failed < failed)
+			fiber_cond_wait(&gc.persist_cond);
+		return gc.persist_completed >= scheduled ? 0 : -1;
+	} else {
+		/*
+		 * Cannot do registration asynchronously - hanging consumer
+		 * can appear in the case of fail.
+		 */
+		struct txn *txn = txn_begin();
+		if (txn == NULL)
+			return -1;
+		txn_set_flags(txn, TXN_FORCE_ASYNC);
+		if (gc_consumer_persist(consumer) != 0) {
+			txn_abort(txn);
+			return -1;
+		}
+		int rc = txn_commit(txn);
+		if (rc == 0)
+			consumer->is_persistent = true;
+		return rc;
+	}
+}
+
+int
+gc_erase_consumer(const struct tt_uuid *uuid)
+{
+	if (!gc_has_persistent_consumers()) {
+		struct gc_consumer *consumer = gc_consumer_by_uuid(uuid);
+		if (consumer != NULL) {
+			consumer->is_persistent = false;
+			if (consumer->ref_count == 0)
+				gc_consumer_unregister_impl(consumer);
+		}
+		return 0;
+	}
+	return boxk(IPROTO_DELETE, BOX_GC_CONSUMERS_ID, "[%s]",
+		    tt_uuid_str(uuid));
```

Comment:
IMO updating in-memory consumer state from on-replace triggers complicates the implementation. What about reading the space directly on recovery, without using on-replace triggers? Then we wouldn't have any branching here - we'd remove the entry from the system space and clear the in-memory state directly.

Reply:
Thanks, I dropped the trigger and read the space directly on recovery.